File: framework/lib/margin.dart
part of uxml;

/**
 * This class represents elements margins and holds the size of its four
 * sides.
 *
 * @author ferhat@ (Ferhat Buyukkokten)
 */
class Margin {
  /** Empty margins */
  static Margin EMPTY;

  /** Sets or returns left margin value */
  num left = 0.0;

  /** Sets or returns top margin value */
  num top = 0.0;

  /** Sets or returns right margin value */
  num right = 0.0;

  /** Sets or returns bottom margin value */
  num bottom = 0.0;

  /**
   * Constructs an immutable rectangle object with optional initialization
   * of its origin and size.
   */
  Margin(this.left, this.top, this.right, this.bottom) {
  }

  /**
   * Constructs margin object with four equals margins.
   */
  Margin.uniform(num size) {
    left = size;
    top = size;
    right = size;
    bottom = size;
  }

  /**
   * Constructs margin object with top and left margins.
   */
  Margin.topLeft(this.top, this.left) {
  }

  /**
   * Checks equality.
   */
  bool equals(Margin compareMargin) {
    return (compareMargin.left == left) &&
        (compareMargin.top == top) &&
        (compareMargin.right == right) &&
        (compareMargin.bottom == bottom);
   }

  /**
   * Grows a rectangle by margin amount.
   */
  Rect grow(Rect rect) {
    return new Rect(rect.x, rect.y, rect.width + left + right,
        rect.height + top + bottom);
  }

  /**
   * Shrinks a rectangle by margin amount.
   */
  Rect shrink(Rect rect) {
    return new Rect(rect.x, rect.y, rect.width - left - right,
        rect.height - top - bottom);
  }

  /**
   * Inflates a rectangle by margin amount.
   */
  Rect inflate(Rect rect) {
    return new Rect(rect.x - left, rect.y - top, rect.width +
        left + right, rect.height + top + bottom);
  }

  /**
   * Deflates a rectangle by margin amount.
   */
  Rect deflate(Rect rect) {
    return new Rect(rect.x + left, rect.y + top, rect.width -
        left - right, rect.height - top - bottom);
  }

  /**
   * Clones object.
   */
  Margin clone() {
    return new Margin(left, top, right, bottom);
  }

  static void initialize() {
    EMPTY = new Margin(0.0, 0.0, 0.0, 0.0);
  }
}

File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/Margin.java
package com.hello.uxml.tools.framework;

/**
 * Represents immutable element margins.
 *
 * @author ferhat
 *
 */
public class Margin{

  private double left;
  private double top;
  private double right;
  private double bottom;

  /**
   * Empty margin object.
   */
  public static final Margin EMPTY = new Margin(0, 0, 0, 0);

  /**
   * Constructor.
   */
  public Margin(double left, double top, double right, double bottom) {
    this.left = left;
    this.top = top;
    this.right = right;
    this.bottom = bottom;
  }

  /**
   * Returns left margin.
   */
  public double getLeft() {
    return left;
  }

  /**
   * Returns top margin.
   */
  public double getTop() {
    return top;
  }

  /**
   * Returns right margin.
   */
  public double getRight() {
    return right;
  }

  /**
   * Returns bottom margin.
   */
  public double getBottom() {
    return bottom;
  }

  /**
   * Returns total left+right margin
   */
  public double getWidth() {
    return left + right;
  }

  /**
   * Returns total top+bottom margin
   */
  public double getHeight() {
    return top + bottom;
  }

  @Override
  public String toString() {
    // Not using String.format due to GWT
    StringBuilder sb = new StringBuilder();
    sb.append("Margin(");
    sb.append(left);
    sb.append(",");
    sb.append(top);
    sb.append(",");
    sb.append(right);
    sb.append(",");
    sb.append(bottom);
    sb.append(")");
    return sb.toString();
  }
}
